### ISSUE TITLE
How do I remove multiple documents using the HTTP API (incorrect documentation/non standard http request)

### ISSUE THREAD

**Issue by mijgame** [arangodb/arangodb#11573]
## My Environment

* __ArangoDB Version__:        3.6.3
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single server
* __Deployment Strategy__:   Docker (arangodb:latest)
* __Configuration__:               Very standard out-of-the-box
* __Infrastructure__:               Locally
* __Operating System__:        Ubuntu 20.04 (Docker host0
* __Total RAM in your machine__:        32Gb
* __Disks in use__:        SSD
* __Used Package__:           N/A

## Component, Query & Data

__Affected feature__:
HTTP API.

__Problem__:
I am trying to delete multiple documents using the HTTP API. I am encountering multiple problems with this:
 1. The examples in the documentation on Swagger/on the website (https://www.arangodb.com/docs/stable/http/document-working-with-documents.html#removes-multiple-documents) are not correct. They are for single document deletions.
2. The textual documentation states that I have to provide a JSON body for the DELETE request. While this is not strictly disallowed by the HTTP spec, a lot of clients don't allow bodies in DELETE requests (such as the .NET Core 3.1  C# HttpClient I am using, see https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request).

Could the documentation be fixed, and can you shed some light on how I should remove multiple documents?

Thanks in advance.

**Comment by Simran-B** [comment]
Hi @mijgame,

### Incorrect Swagger examples

Thanks for letting us know. The examples are indeed incorrect. I'm verifying a new set of examples for multi-delete and will publish them later today.

### HTTP GET/DELETE requests with bodies

**TL;DR**
If .NET Core 3.1's C# HttpClient doesn't support a body for DELETE requests, then it is not standard compliant in my opinion. My reasoning:

Under _4.3 Message Body_ of [RFC 2616](https://tools.ietf.org/html/rfc2616) it is said that:

> A message-body MUST NOT be included in a request if the specification of the request method (section 5.1.1) does not allow sending an entity-body in requests. A server SHOULD read and forward a message-body on any request; if the request method does not include defined semantics for an entity-body, then the message-body SHOULD be ignored when handling the request.

This is somewhat ambiguous. It forbids bodies for certain methods, but also says that they should be read & forwarded as long as the semantics are defined. Anyway, if we take a look at _5.1.1 Method_ then we don't find any statement about methods for which bodies are disallowed. It tells us that the semantics for methods are specified under _9 Method Definitions_ however:

> [...] if the above methods are implemented, they MUST be implemented with the same semantics as those specified in section 9.

Let us first check _9.3 GET_:

> The GET method means retrieve whatever information (in the form of an entity) is identified by the Request-URI.

And _9.7 DELETE_:

> The DELETE method requests that the origin server delete the resource identified by the Request-URI.

This implicitly states that a potential body must not play a role in the identification of the resource in GET and DELETE requests. The target must be solely identified by the request URL. We expect the body to contain an array of selectors for multi-document API requests in ArangoDB however, so for now we could say that ArangoDB isn't following the standard.

But! RFC 2616 is from 1999 and obsoleted by [RFC 7231](https://tools.ietf.org/html/rfc7231) from 2014. Let's check the current spec.

In _4.3.1.  GET_ under _4.3.  Method Definitions_ it does no longer say that the Request-URI has to identify the information you want to retrieve:

> The GET method requests transfer of a current selected representation for the target resource.
> 
> The HTTP interface for a resource is just as likely to be implemented as a tree of content objects, a programmatic view on various database records, or a gateway to other information systems.  Even when the URI mapping mechanism is tied to a file system, an origin server might be configured to execute the files with the request as input and send the output as the representation rather than transfer the files directly.  Regardless, only the origin server needs to know how each of its resource identifiers corresponds to an implementation and how each implementation manages to select and send a current representation of the target resource in a response to GET.

_3.2.  Representation Data_:

> The representation data associated with an HTTP message is either provided as the payload body of the message or referred to by the message semantics and the effective request URI. 

Bodies in GET and DELETE requests do still not have defined semantics, and they acknowledge that old implementations may not support them:

> A payload within a GET/DELETE request message has no defined semantics; sending a payload body on a GET/DELETE request might cause some existing implementations to reject the request.

The statement from RFC 2616...

> if the request method does not include defined semantics for an entity-body, then the message-body SHOULD be ignored when handling the request

... was deleted however. In addition, [RFC 7230](https://tools.ietf.org/html/rfc7230) specifies the following under _3.3.  Message Body_:

> The presence of a message body in a request is signaled by a Content-Length or Transfer-Encoding header field.  Request message framing is **independent of method semantics, even if the method does not define any use for a message body**.

The only exception is _4.3.8.  TRACE_:

> A client MUST NOT send a message body in a TRACE request.

**Conclusion:**
RFC 7231 explicitly permits message bodies for GET and DELETE, and my interpretation of RFC 7230 is that it is even required to support them (in the sense that the bodies are not to be ignored or omitted). The .NET Core implementation might be RFC 2616 compliant, but the spec is outdated (6 years and counting).

---

Multi document GET requests shouldn't be a problem via ArangoDB's HTTP API, because the route doesn't use `GET`:

    PUT /_api/document/{collection}?onlyget=true

Beware however, if you omit the query parameter `onlyget` or set it to `false` then this will be a replace operation!

There is no such workaround for multi document DELETE unfortunately. There are other workarounds however:

- Use AQL (`/_api/cursor`), for example:
  Query: `FOR key IN @keys REMOVE key IN @@coll RETURN KEEP(OLD, "_key", "_id", "_rev")`
  BindVars: `{ "keys": ["1", "2"], "@coll": "products" }`
- Combine single document deletion API requests with [Batch requests](https://www.arangodb.com/docs/stable/http/batch-request.html)
- Use [Stream Transactions](https://www.arangodb.com/docs/stable/http/transaction-stream-transaction.html)

**Comment by Simran-B** [comment]
The new examples are available here:
https://www.arangodb.com/docs/devel/http/document-working-with-documents.html#removes-multiple-documents

**Comment by mijgame** [comment]
Thanks you for the quick and extensive response!

I agree with you that the HttpClient is not standards compliant here. The cursor workaround is what we are using right now, which works fine. We were already using the multiple get, so I wanted to keep these calls consistent.

Am I correct in saying that the multiple GET request is actually inconsistent with the other calls? Per the spec the multiple GET and multiple DELETE operations should work in the same way (using a body), instead of a PUT request with `onlyget=true`?

Maybe there should be a PUT request with `onlydelete=true` or something like that, to keep consistent while also allowing older/not strictly spec compliant clients to delete multiple documents.

**Comment by Simran-B** [comment]
You are welcome!

Multi-get is inconsistent with the rest of the API, that's right. The reason to implement it like this was obviously because of implementations which don't allow GET with a body, in particular firewalls and load-balancers I believe. Why this wasn't done for multi-delete which seems to have the same issues is beyond me. Maybe an oversight on our end, or perhaps because DELETE with a body doesn't cause as much problems in practice as it generally receives less attention among the HTTP methods. It is non-cachable IIRC, which could also have something to do with it.

Adding an `onlydelete` query parameter would be logical and should be backward-compatible. I will thus mark this issue as a feature request. For 4.0 we might reconsider the API design. My preference would be to use GET + body and DELETE + body as default, but offer alternatives similar to `onlyget`. The POST method would be more suitable than PUT in my opinion (avoid accidental replace operations that were supposed to be reads).
